source/dub/dependencyresolver.d

```diff
@@ -54,236 +54,149 @@ class DependencyResolver(CONFIGS, CONFIG) {
-					else cs[i] = p ~ " [no config]";
+	private void constrainDependencies(TreeNode n, TreeNodes[] dependencies, size_t depidx, ref PackageConfigs[string] packs, ref long loop_counter)
+	{
+		if (depidx >= dependencies.length) return;
+
+		assert (--loop_counter > 0,
+			"The dependency resolution process is taking too long. The"
+			~ " dependency graph is likely hitting a pathological case in"
+			~ " the resolution algorithm. Please file a bug report at"
+			~ " https://github.com/dlang/dub/issues and mention the package"
+			~ " recipe that reproduces this error.");
+
+		auto dep = &dependencies[depidx];
+		auto depbase = dep.pack.basePackage;
+		auto dp = &packs[depbase];
+
+		Exception first_err;
+
+		// try each configuration/version of the current dependency
+		foreach (i, c; dp.configs) {
+			if (dp.included[i]) {
+				auto subpacks = clone(packs);
+				foreach (j, ref inc; subpacks[depbase].included) inc = j == i;
+				try {
+					constrainRec(TreeNode(dep.pack, c), subpacks, loop_counter);
+					constrainDependencies(TreeNode(dep.pack, c), dependencies, depidx+1, subpacks, loop_counter);
+					// if a branch succeeded, replace the current package info
+					// with the one from the branch and return
+					packs = subpacks;
+					return;
+				} catch (Exception e) {
+					if (!first_err) first_err = e;
 				}
-				return cs.join(", ");
-			}());
-
-			if (conflict_index < 0) {
-				CONFIG[string] ret;
-				foreach (p, i; package_indices)
-					if (configs[i].allConfigs.length) {
-						auto cfg = configs[i].allConfigs[config_indices[i]];
-						if (cfg != CONFIG.invalid) ret[p] = cfg;
-					}
-				logDebug("Resolved dependencies before optional-purge: %s", ret.byKey.map!(k => k~" "~ret[k].to!string));
-				purgeOptionalDependencies(root, ret);
-				logDebug("Resolved dependencies after optional-purge: %s", ret.byKey.map!(k => k~" "~ret[k].to!string));
-				return ret;
-			}
-
-			// find the next combination of configurations
-			foreach_reverse (pi, ref i; config_indices) {
-				if (pi > conflict_index) i = 0;
-				else if (++i >= configs[pi].allConfigs.length) i = 0;
-				else break;
-			}
-			if (config_indices.all!"a==0") {
-				if (throw_on_failure) throw new Exception(format("Could not find a valid dependency tree configuration: %s", first_error.get));
-				else return null;
 			}
 		}
-	}
 
-	protected abstract CONFIG[] getAllConfigs(string pack);
-	protected abstract CONFIG[] getSpecificConfigs(string pack, TreeNodes nodes);
-	protected abstract TreeNodes[] getChildren(TreeNode node);
-	protected abstract bool matches(CONFIGS configs, CONFIG config);
+		// ignore unsatisfyable optional dependencies
+		if (dep.depType != DependencyType.required)
+			return;
+
+		// report the first error encountered to the user
+		if (first_err) throw first_err;
+
+		// should have thrown in constrainRec before reaching this
+		assert(false, "Got no configuration for dependency!?");
```

Comment:
just in case this actually gets triggered, including which package triggered this would probably be a good idea.